Add blog post on Best Practices of Securing Production Debugging in Kubernetes (#54582)
kubernetes/website: content/en/blog/_posts/2026/securing-production-debugging-in-kubernetes.md

The goal is to use short-lived, identity-bound credentials that clearly tie a session to a real person and expire quickly. These credentials can include the user’s identity and the scope of what they’re allowed to do. They’re typically signed using a private key that stays with the engineer, such as a hardware-backed key (for example, a YubiKey), so they can not be forged without access to that key.

You can implement this with Kubernetes-native auth (for example, client certificates or an OIDC-based flow), or have the access broker from the previous section issue short-lived credentials on the user’s behalf. In many setups, Kubernetes still uses RBAC to enforce permissions based on the authenticated identity and groups/claims. If you use an access broker, it can also encode additional scope constraints in the credential and enforce them during the session, such as which cluster or namespace the session applies to and which actions (or approved commands) are allowed against pods or nodes. In either case, the credentials should be signed by a certificate authority (CA), and that CA should be rotated on a regular schedule (for example, quarterly) to limit long-term risk.

> ### Option A: short-lived OIDC tokens

Reviewer: I am dubious about this. I think the heading misleads. Maybe you as the author think people should use OIDC, or perhaps you are telling readers that tokens for authn must be OIDC.

The reader can't tell.

This post offers my recommendations for good practices applicable to existing Kubernetes environments with minimal tooling changes:

- Least privilege with RBAC 
- Short-lived, identity-bound credentials
- An SSH-style handshake model for cloud-native debugging 

A good architecture for securing production debugging workflows is to use a just-in-time secure shell gateway
(often deployed as an on demand pod in the cluster).
It acts as an SSH-style “front door” that makes temporary access actually temporary. You can 
authenticate with short-lived, identity-bound credentials, establish a session to the gateway,
and the gateway uses the Kubernetes API and RBAC to control what they can do such as `pods/log`, `pods/exec`, and `pods/portforward`.
Sessions expire automatically, and both the gateway logs and Kubernetes audit logs capture who accessed what and when without shared bastion accounts or long-lived keys.

## 1) Using an access broker on top of Kubernetes RBAC

> RBAC is about control: who can do what. Many Kubernetes environments rely primarily on RBAC for authorization, although Kubernetes also supports other authorization modes such as Webhook authorization. You can enforce it directly with Kubernetes RBAC, or put an access broker/gateway in front of the cluster that still relies on Kubernetes permissions under the hood. That access broker is useful for decisions RBAC does not cover well, like whether a request should be auto-approved or require manual approval, can a user run a command or not, and what kind of commands are allowed for sessions. This access broker will also be responsible for managing group membership, so permissions are granted at the group level rather than to individual users. Kubernetes RBAC can decide whether someone is allowed to use actions like `pods/exec`, but it can not limit which commands run inside an exec session. An access broker can add those extra guardrails, while RBAC remains the source of truth for what the Kubernetes API will allow and at what scope.

Reviewer: Rewordings and wrapping.
```markdown
RBAC is about control: who can do what. Many Kubernetes environments rely primarily on RBAC for authorization,
although Kubernetes also supports other authorization modes such as Webhook authorization.
You can enforce it directly with Kubernetes RBAC, or put an access broker/gateway in front of the cluster
that still relies on Kubernetes permissions under the hood.
That access broker is useful for decisions RBAC does not cover well, like whether a request should be auto-approved or require manual approval,
whether a user particular user can run a command or not, and what kind of commands are allowed for sessions.
This access broker is typically also be responsible for managing group membership, so that permissions are granted at the group level
rather than to individual users.
Kubernetes RBAC can control whether someone is allowed to use actions like `pods/exec`, but it can not (by itself) limit which commands run inside an exec session.
An access broker can add those extra guardrails, while RBAC remains the source of truth for what the Kubernetes API will allow and at what scope.
```
Also consider hyperlinking to https://kubernetes.io/docs/reference/access-authn-authz/rbac/ and to https://kubernetes.io/docs/reference/access-authn-authz/webhook/

Only take what you agree with here; the blog article needs to be in **your** words. But if you do see a change and can see why I've suggested it and you think it will make the article make more sense, you should accept it.

draft: true
slug: securing-production-debugging-in-kubernetes
author: >
  Shridivya Sharma
---

During production debugging, the fastest route is often broad access such as `cluster-admin` (very high-level admin access to a Kubernetes cluster, similar to how Contributor in Azure or PowerUser-style access in AWS gives broad access to resources), shared bastions/jump boxes, or long-lived SSH keys. It works in the moment, but it comes with two common problems: auditing becomes difficult, and temporary exceptions have a way of becoming routine.

This post offers my recommendations for good practices applicable to existing Kubernetes environments with minimal tooling changes:

- Least privilege with RBAC 
- Short-lived, identity-bound credentials
- An SSH-style handshake model for cloud-native debugging 

A good architecture for securing production debugging workflows is to use a just-in-time secure shell gateway
(often deployed as an on demand pod in the cluster).
It acts as an SSH-style “front door” that makes temporary access actually temporary. You can 
authenticate with short-lived, identity-bound credentials, establish a session to the gateway,

> and the gateway uses the Kubernetes API and RBAC to control what they can do such as `pods/log`, `pods/exec`, and `pods/portforward`.

Reviewer 2: ```suggestion
It acts as an SSH-style “front door” that makes temporary access actually temporary. You can
authenticate with short-lived, identity-bound credentials, establish a session to the gateway,
and the gateway uses the Kubernetes API and RBAC to control what they can do, such as `pods/log`, `pods/exec`, and `pods/portforward`.
```

Sessions expire automatically, and both the gateway logs and Kubernetes audit logs capture who accessed what and when without shared bastion accounts or long-lived keys.

## 1) Using an access broker on top of Kubernetes RBAC

> RBAC is about control: who can do what. Many Kubernetes environments rely primarily on RBAC for authorization, although Kubernetes also supports other authorization modes such as Webhook authorization. You can enforce it directly with Kubernetes RBAC, or put an access broker/gateway in front of the cluster that still relies on Kubernetes permissions under the hood. That access broker is useful for decisions RBAC does not cover well, like whether a request should be auto-approved or require manual approval, can a user run a command or not, and what kind of commands are allowed for sessions. This access broker will also be responsible for managing group membership, so permissions are granted at the group level rather than to individual users. Kubernetes RBAC can decide whether someone is allowed to use actions like `pods/exec`, but it can not limit which commands run inside an exec session. An access broker can add those extra guardrails, while RBAC remains the source of truth for what the Kubernetes API will allow and at what scope.

Reviewer 2: We may want to consider splitting this paragraph, according to [the style guide we try to keep them under 6 sentences](https://kubernetes.io/docs/contribute/style/style-guide/#paragraphs).

> RBAC is about control: who can do what. Many Kubernetes environments rely primarily on RBAC for authorization, although Kubernetes also supports other authorization modes such as Webhook authorization. You can enforce it directly with Kubernetes RBAC, or put an access broker/gateway in front of the cluster that still relies on Kubernetes permissions under the hood. That access broker is useful for decisions RBAC does not cover well, like whether a request should be auto-approved or require manual approval, can a user run a command or not, and what kind of commands are allowed for sessions. This access broker will also be responsible for managing group membership, so permissions are granted at the group level rather than to individual users. Kubernetes RBAC can decide whether someone is allowed to use actions like `pods/exec`, but it can not limit which commands run inside an exec session. An access broker can add those extra guardrails, while RBAC remains the source of truth for what the Kubernetes API will allow and at what scope.

Author: Done.

You can implement this with Kubernetes-native authentication (for example, client certificates or an OIDC-based flow),
or have the access broker from the previous section issue short-lived credentials on the user’s behalf.
In many setups, Kubernetes still uses RBAC to enforce permissions based on the authenticated identity and groups/claims.
If you use an access broker, it can also encode additional scope constraints in the credential and enforce them during the session, such as which cluster or namespace the session applies to and which actions
(or approved commands) are allowed against pods or nodes.
In either case, the credentials should be signed by a certificate authority (CA), and that CA should be rotated on a regular schedule (for example, quarterly) to limit long-term risk.

### Option A: short-lived OIDC tokens

> A lot of managed Kubernetes clusters already give you short-lived tokens. The main thing is to make sure your kubeconfig refreshes them automatically  instead of copying a long-lived token into the file.

Reviewer 2: ```suggestion
A lot of managed Kubernetes clusters already give you short-lived tokens. The main thing is to make sure your kubeconfig refreshes them automatically instead of copying a long-lived token into the file.
```

### Option A: short-lived OIDC tokens

A lot of managed Kubernetes clusters already give you short-lived tokens. The main thing is to make sure your kubeconfig refreshes them automatically instead of copying a long-lived token into the file.

For example:

```yaml
users:
- name: oncall
  user:
    exec:
      apiVersion: client.authentication.k8s.io/v1
      command: cred-helper
      args: ["--cluster=prod", "--ttl=30m"]
```

### Option B: Short-lived client certificates (X.509)

If your API server (or your access broker from the previous section) is set up to trust a client CA, you can use short-lived client certificates for debugging access. The idea is:

* The private key is created and kept under the engineer’s machine (ideally hardware-backed, like a non-exportable key in a YubiKey/PIV token)
* A short-lived certificate is issued (often via the

>   [CertificateSigningRequest](/docs/reference/access-authn-authz/certificate-signing-requests/#certificate-signing-requests) API, or your access broker from the previous session with a TTL)

Reviewer 2: do you mean "previous section" here, like above?

> This is pretty straightforward to operationalize with the Kubernetes CertificateSigningRequest API.

Reviewer 2: ```suggestion
This is straightforward to operationalize with the Kubernetes CertificateSigningRequest API.
```

    resources: ["pods", "pods/log"]
    verbs: ["get", "list", "watch"]
  - apiGroups: [""]
    resources: ["pods/exec"]
    verbs: ["create"]
---
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: jit-debug
  namespace: <namespace>
subjects:
  - kind: Group
    name: jit:oncall:<namespace>   # mapped from the short-lived credential (cert/OIDC)
    apiGroup: rbac.authorization.k8s.io
roleRef:
  kind: Role
  name: jit-debug
  apiGroup: rbac.authorization.k8s.io
```
These RBAC objects, and the rules they define, allow debugging only within the specified namespace; attempts to access other namespaces are not allowed.

### Example: Cluster-scoped role binding

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: jit-cluster-read
rules:
  - apiGroups: [""]
    resources: ["nodes", "namespaces"]
    verbs: ["get", "list", "watch"]
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: jit-cluster-read
subjects:
  - kind: Group
    name: jit:oncall:cluster
    apiGroup: rbac.authorization.k8s.io
roleRef:
  kind: ClusterRole
  name: jit-cluster-read
  apiGroup: rbac.authorization.k8s.io

> 

Reviewer 2: remove blank line before end of code fence.

Author: done

```

> This RBAC policy grants cluster-wide read access (for example, to nodes and namespaces) and should be used only for workflows that truly require cluster-scoped resources.

Reviewer: ```suggestion
These RBAC rules grant cluster-wide read access (for example, to nodes and namespaces) and should be used only for workflows that truly require cluster-scoped resources.
```
